`test_batch_renormalization` is flaky

This is part of https://github.com/chainer/chainer/issues/6903 effort.

- Failure example:  https://ci.preferred.jp/chainer.py37/2214/
- Failure test case: `BatchRenormalizationTest_param_21_{dtype=float64, ndim=2, test=False}.test_forward_multi_gpu`
- Tolerance:
```
E   Not equal to tolerance rtol=0.001, atol=0.0001
E   
E   Mismatch: 6.67%
E   Max absolute difference: 0.00052644
E   Max relative difference: 0.00564455
```

https://jenkins.preferred.jp/job/chainer/job/chainer_pr/1839/TEST=chainer-py3,label=mn1-p100/console

>`FAIL tests/chainer_tests/links_tests/normalization_tests/test_batch_renormalization.py::BatchRenormalizationTest_param_11_{dtype=float32, ndim=1, test=False}::test_forward_gpu`

```
18:07:41 _ BatchRenormalizationTest_param_11_{dtype=float32, ndim=1, test=False}.test_forward_gpu _
18:07:41 chainer/testing/parameterized.py:89: in new_method
18:07:41     utils._raise_from(e.__class__, s.getvalue(), e)
18:07:41 chainer/utils/__init__.py:106: in _raise_from
18:07:41     six.raise_from(new_exc.with_traceback(orig_exc.__traceback__), None)
18:07:41 chainer/testing/parameterized.py:78: in new_method
18:07:41     return base_method(self, *args, **kwargs)
18:07:41 chainer/testing/condition.py:80: in wrapper
18:07:41     fail()
18:07:41 chainer/testing/condition.py:51: in fail
18:07:41     instance.fail(msg)
18:07:41 E   AssertionError: Parameterized test failed.
18:07:41 E   
18:07:41 E   Base test method: BatchRenormalizationTest.test_forward_gpu
18:07:41 E   Test parameters:
18:07:41 E     dtype: <class 'numpy.float32'>
18:07:41 E     ndim: 1
18:07:41 E     test: False
18:07:41 E   
18:07:41 E   
18:07:41 E   (caused by)
18:07:41 E   AssertionError: 
18:07:41 E   Fail: 3, Success: 0
18:07:41 E   
18:07:41 E   The first error message:
18:07:41 E   Traceback (most recent call last):
18:07:41 E     File "/work/chainer/chainer/testing/array.py", line 27, in assert_allclose
18:07:41 E       x, y, atol=atol, rtol=rtol, verbose=verbose)
18:07:41 E     File "/opt/pyenv/versions/3.7.1/lib/python3.7/site-packages/numpy/testing/_private/utils.py", line 1510, in assert_allclose
18:07:41 E       verbose=verbose, header=header, equal_nan=equal_nan)
18:07:41 E     File "/opt/pyenv/versions/3.7.1/lib/python3.7/site-packages/numpy/testing/_private/utils.py", line 836, in assert_array_compare
18:07:41 E       raise AssertionError(msg)
18:07:41 E   AssertionError: 
18:07:41 E   Not equal to tolerance rtol=0.001, atol=0.0001
18:07:41 E   
18:07:41 E   Mismatch: 3.33%
18:07:41 E   Max absolute difference: 0.00054932
18:07:41 E   Max relative difference: 0.00133648
18:07:41 E    x: array([[[-2.084918,  0.409576],
18:07:41 E           [-0.363258, -0.557226],
18:07:41 E           [ 1.296712,  5.226429]],...
18:07:41 E    y: array([[[-2.08488 ,  0.409581],
18:07:41 E           [-0.363256, -0.557228],
18:07:41 E           [ 1.296163,  5.22588 ]],...
18:07:41 E   
18:07:41 E   During handling of the above exception, another exception occurred:
18:07:41 E   
18:07:41 E   Traceback (most recent call last):
18:07:41 E     File "/work/chainer/chainer/testing/condition.py", line 60, in <lambda>
18:07:41 E       lambda: f(ins, *args[1:], **kwargs),
18:07:41 E     File "/work/chainer/tests/chainer_tests/links_tests/normalization_tests/test_batch_renormalization.py", line 107, in test_forward_gpu
18:07:41 E       self.check_forward(cuda.to_gpu(self.x))
18:07:41 E     File "/work/chainer/tests/chainer_tests/links_tests/normalization_tests/test_batch_renormalization.py", line 97, in check_forward
18:07:41 E       y_expect, y.data, **self.check_forward_optionss)
18:07:41 E     File "/work/chainer/chainer/testing/array.py", line 68, in assert_allclose
18:07:41 E       raise AssertionError(f.getvalue())
18:07:41 E   AssertionError: 
18:07:41 E   Not equal to tolerance rtol=0.001, atol=0.0001
18:07:41 E   
18:07:41 E   Mismatch: 3.33%
18:07:41 E   Max absolute difference: 0.00054932
18:07:41 E   Max relative difference: 0.00133648
18:07:41 E    x: array([[[-2.084918,  0.409576],
18:07:41 E           [-0.363258, -0.557226],
18:07:41 E           [ 1.296712,  5.226429]],...
18:07:41 E    y: array([[[-2.08488 ,  0.409581],
18:07:41 E           [-0.363256, -0.557228],
18:07:41 E           [ 1.296163,  5.22588 ]],...
18:07:41 E   
18:07:41 E   assert_allclose failed: 
18:07:41 E     shape: (5, 3, 2) (5, 3, 2)
18:07:41 E     dtype: float32 float32
18:07:41 E     i: (2, 2, 1)
18:07:41 E     x[i]: 0.41107428073883057
18:07:41 E     y[i]: 0.4105256199836731
18:07:41 E     relative error[i]: 0.0013364836340770125
18:07:41 E     absolute error[i]: 0.0005486607551574707
18:07:41 E     relative tolerance * |y[i]|: 0.0004105256199836731
18:07:41 E     absolute tolerance: 0.0001
18:07:41 E     total tolerance: 0.0005105256199836731
18:07:41 E   x: [[[-2.0849183   0.40957627]
18:07:41 E        [-0.36325765 -0.55722564]
18:07:41 E        [ 1.2967116   5.226429  ]]
18:07:41 E   
18:07:41 E       [[-0.7658379   0.07419053]
18:07:41 E        [ 0.7500052  -0.47575486]
18:07:41 E        [ 1.0481513   1.7667365 ]]
18:07:41 E   
18:07:41 E       [[ 0.08743405 -0.93879426]
18:07:41 E        [ 0.2873434  -0.38592124]
18:07:41 E        [ 4.483272    0.41107428]]
18:07:41 E   
18:07:41 E       [[-1.8458759  -0.68705523]
18:07:41 E        [ 0.304021    0.8170537 ]
18:07:41 E        [ 1.0541122   2.3461757 ]]
18:07:41 E   
18:07:41 E       [[-2.0749302  -0.50755703]
18:07:41 E        [-0.01045106 -0.35363   ]
18:07:41 E        [ 0.8636067   5.171902  ]]]
18:07:41 E   y: [[[-2.08488     0.40958092]
18:07:41 E        [-0.3632562  -0.5572276 ]
18:07:41 E        [ 1.296163    5.2258797 ]]
18:07:41 E   
18:07:41 E       [[-0.76581746  0.07419983]
18:07:41 E        [ 0.7500259  -0.47575536]
18:07:41 E        [ 1.0476025   1.7661879 ]]
18:07:41 E   
18:07:41 E       [[ 0.08744311 -0.9387715 ]
18:07:41 E        [ 0.28735608 -0.38592023]
18:07:41 E        [ 4.482723    0.41052562]]
18:07:41 E   
18:07:41 E       [[-1.845841   -0.6870359 ]
18:07:41 E        [ 0.30403394  0.81707555]
18:07:41 E        [ 1.0535635   2.345627  ]]
18:07:41 E   
18:07:41 E       [[-2.0748923  -0.50754005]
18:07:41 E        [-0.01044352 -0.35362846]
18:07:41 E        [ 0.86305803  5.1713533 ]]]
```

https://travis-ci.org/chainer/chainer/jobs/574189720

>`FAIL tests/chainer_tests/links_tests/normalization_tests/test_batch_renormalization.py::BatchRenormalizationTest_param_11_{ndim=1, test=False, dtype=float32}::test_forward_cpu`

```
E   assert_allclose failed: 
E     shape: (5, 3, 2) (5, 3, 2)
E     dtype: float32 float32
E     i: (1, 1, 0)
E     x[i]: 0.024260878562927246
E     y[i]: 0.024406611919403076
E     relative error[i]: 0.00597106060013175
E     absolute error[i]: 0.00014573335647583008
E     relative tolerance * |y[i]|: 2.4406611919403078e-05
E     absolute tolerance: 0.0001
E     total tolerance: 0.0001244066119194031
```

--- a/tests/chainer_tests/links_tests/normalization_tests/test_batch_renormalization.py
+++ b/tests/chainer_tests/links_tests/normalization_tests/test_batch_renormalization.py
@@ -54,8 +54,6 @@ class BatchRenormalizationTest(unittest.TestCase):
         if self.test:
             self.mean = numpy.random.uniform(-1, 1, (3,)).astype(self.dtype)
             self.var = numpy.random.uniform(0.5, 1, (3,)).astype(self.dtype)
-            self.link.avg_mean[...] = self.mean
-            self.link.avg_var[...] = self.var
             self.running_mean = self.mean
             self.running_var = self.var
         else:
@@ -63,16 +61,19 @@ class BatchRenormalizationTest(unittest.TestCase):
             self.var = self.x.var(axis=self.aggr_axes)
             # Need to add some noise to running_mean and running_var,
             # otherwise we will always get r=1, d=0
-            self.running_mean = self.mean + numpy.random.uniform(
-                -1, 1, self.mean.shape).astype(self.dtype)
-            self.running_var = numpy.abs(self.var + numpy.random.uniform(
-                -1, 1, self.var.shape).astype(self.dtype))
-            self.link.avg_mean[...] = self.running_mean
-            self.link.avg_var[...] = self.running_var
+            # Note that numpy.exp(3) > rmax ** 2 and 7 > dmax
+            self.running_var = self.var * numpy.exp(
+                numpy.random.uniform(-3, 3, self.var.shape)).astype(self.dtype)
+            self.running_mean = self.mean + (
+                (numpy.sqrt(self.running_var) + 0.1)
+                * numpy.random.uniform(-7, 7, self.mean.shape)
+            ).astype(self.dtype)
+        self.link.avg_mean[...] = self.running_mean
+        self.link.avg_var[...] = self.running_var
         self.check_forward_optionss = {'atol': 1e-4, 'rtol': 1e-3}
         self.check_backward_optionss = {'atol': 1e-4, 'rtol': 1e-3}
         if self.dtype == numpy.float16:
-            self.check_forward_optionss = {'atol': 1e-3, 'rtol': 1e-2}
+            self.check_forward_optionss = {'atol': 1e-2, 'rtol': 5e-3}
             self.check_backward_optionss = {'atol': 5e-1, 'rtol': 1e-1}
 
     def check_forward(self, x_data):

--- a/tests/chainer_tests/links_tests/normalization_tests/test_batch_renormalization.py
+++ b/tests/chainer_tests/links_tests/normalization_tests/test_batch_renormalization.py
@@ -8,17 +8,14 @@ from chainer.backends import cuda
 from chainer import links
 from chainer import testing
 from chainer.testing import attr
-from chainer.testing import condition
 
 
-def _batch_renormalization(expander, gamma, beta, x, mean, var, eps, test,
+def _batch_renormalization(expander, gamma, beta, x, mean, std, test,
                            r, d):
     mean = mean[expander]
+    std = std[expander]
     if test:
-        std = numpy.sqrt(var[expander])
         r, d = 1, 0
-    else:
-        std = numpy.sqrt(var[expander] + eps)
     y_expect = gamma * ((x - mean) / std * r + d) + beta
     return y_expect
 
@@ -27,6 +24,7 @@ def _batch_renormalization(expander, gamma, beta, x, mean, var, eps, test,
     'test': [True, False],
     'ndim': [0, 1, 2, 3],
     'dtype': [numpy.float16, numpy.float32, numpy.float64],
+    'eps': [2e-5, 0.5],
 })))
 class BatchRenormalizationTest(unittest.TestCase):
 
@@ -37,9 +35,9 @@ class BatchRenormalizationTest(unittest.TestCase):
         self.rmax = self.dtype(3)
         self.dmax = self.dtype(5)
 
-        self.link = links.BatchRenormalization(3, rmax=self.rmax,
-                                               dmax=self.dmax,
-                                               dtype=self.dtype)
+        self.link = links.BatchRenormalization(
+            3, rmax=self.rmax, dmax=self.dmax,
+            dtype=self.dtype, eps=self.eps)
         gamma = self.link.gamma.data
         gamma[...] = numpy.random.uniform(.5, 1, gamma.shape)
         beta = self.link.beta.data
@@ -83,31 +81,28 @@ class BatchRenormalizationTest(unittest.TestCase):
             y = self.link(x)
             self.assertEqual(y.data.dtype, self.dtype)
 
-        sigma_batch = numpy.sqrt(self.var)
-        running_sigma = numpy.sqrt(self.running_var)
+        sigma_batch = numpy.sqrt(self.var + self.eps)
+        running_sigma = numpy.sqrt(self.running_var + self.eps)
         r = numpy.clip(sigma_batch / running_sigma, 1.0 / self.rmax, self.rmax)
         d = numpy.clip((self.mean - self.running_mean) / running_sigma,
                        -self.dmax, self.dmax)
         y_expect = _batch_renormalization(
             self.expander, self.gamma, self.beta, self.x, self.mean,
-            self.var, self.link.eps, self.test,
+            sigma_batch, self.test,
             r[self.expander], d[self.expander])
 
         testing.assert_allclose(
             y_expect, y.data, **self.check_forward_optionss)
 
-    @condition.retry(3)
     def test_forward_cpu(self):
         self.check_forward(self.x)
 
     @attr.gpu
-    @condition.retry(3)
     def test_forward_gpu(self):
         self.link.to_gpu()
         self.check_forward(cuda.to_gpu(self.x))
 
     @attr.multi_gpu(2)
-    @condition.retry(3)
     def test_forward_multi_gpu(self):
         with cuda.get_device_from_id(1):
             self.link.to_gpu()

--- a/tests/chainer_tests/links_tests/normalization_tests/test_batch_renormalization.py
+++ b/tests/chainer_tests/links_tests/normalization_tests/test_batch_renormalization.py
@@ -38,9 +38,9 @@ class BatchRenormalizationTest(unittest.TestCase):
         self.link = links.BatchRenormalization(
             3, rmax=self.rmax, dmax=self.dmax,
             dtype=self.dtype, eps=self.eps)
-        gamma = self.link.gamma.data
+        gamma = self.link.gamma.array
         gamma[...] = numpy.random.uniform(.5, 1, gamma.shape)
-        beta = self.link.beta.data
+        beta = self.link.beta.array
         beta[...] = numpy.random.uniform(-1, 1, beta.shape)
         self.link.cleargrads()
 
@@ -80,7 +80,7 @@ class BatchRenormalizationTest(unittest.TestCase):
         with chainer.using_config('train', not self.test):
             x = chainer.Variable(x_data)
             y = self.link(x)
-            self.assertEqual(y.data.dtype, self.dtype)
+            self.assertEqual(y.dtype, self.dtype)
 
         sigma_batch = numpy.sqrt(self.var + self.eps)
         running_sigma = numpy.sqrt(self.running_var + self.eps)
@@ -93,7 +93,7 @@ class BatchRenormalizationTest(unittest.TestCase):
             r[self.expander], d[self.expander])
 
         testing.assert_allclose(
-            y_expect, y.data, **self.check_forward_optionss)
+            y.array, y_expect, **self.check_forward_optionss)
 
     def test_forward_cpu(self):
         self.check_forward(self.x)
@@ -135,15 +135,15 @@ class TestPopulationStatistics(unittest.TestCase):
         x = chainer.Variable(x)
         self.link(x, finetune=True)
         mean = self.x.mean(axis=0)
-        testing.assert_allclose(mean, self.link.avg_mean)
+        testing.assert_allclose(self.link.avg_mean, mean)
         unbiased_var = self.x.var(axis=0) * self.nx / (self.nx - 1)
-        testing.assert_allclose(unbiased_var, self.link.avg_var)
+        testing.assert_allclose(self.link.avg_var, unbiased_var)
 
         y = chainer.Variable(y)
         with chainer.using_config('train', False):
             self.link(y, finetune=True)
-        testing.assert_allclose(mean, self.link.avg_mean)
-        testing.assert_allclose(unbiased_var, self.link.avg_var)
+        testing.assert_allclose(self.link.avg_mean, mean)
+        testing.assert_allclose(self.link.avg_var, unbiased_var)
 
     def test_statistics_cpu(self):
         self.check_statistics(self.x, self.y)
@@ -169,8 +169,8 @@ class TestPopulationStatistics(unittest.TestCase):
         # But the multiplier is ny / (ny - 1) in current implementation
         # these two values are different when nx is not equal to ny.
         unbiased_var = var * self.ny / (self.ny - 1)
-        testing.assert_allclose(mean, self.link.avg_mean)
-        testing.assert_allclose(unbiased_var, self.link.avg_var)
+        testing.assert_allclose(self.link.avg_mean, mean)
+        testing.assert_allclose(self.link.avg_var, unbiased_var)
 
     def test_statistics2_cpu(self):
         self.check_statistics2(self.x, self.y)

